include/EASTL/bonus/ring_buffer.h

```diff
@@ -1505,7 +1505,12 @@ namespace eastl
 	template <typename T, typename Container, typename Allocator>
 	inline bool operator==(const ring_buffer<T, Container, Allocator>& a, const ring_buffer<T, Container, Allocator>& b)
 	{
-		return (a.size() == b.size()) && (a.c == b.c);
+		const typename ring_buffer<T, Container, Allocator>::size_type sizeA = a.size();
+		const typename ring_buffer<T, Container, Allocator>::size_type sizeB = b.size();
+
+		if(sizeA == sizeB)
+			return eastl::equal(a.begin(), a.end(), b.begin());
+		return sizeA < sizeB;
```

Comment:
I don't think this is correct. Shouldn't this unconditionally return `false` if the ring buffers have different sizes?

Reply:
You are right，it is an obvious mistake. 😅 Fixed.

I have found that it may be possible to avoid calling `size()` here and instead use the `eastl::identical` interface for more precise semantics and readability. However, it seems that this interface is not mandated by the standard. Which approach do you think is better?

```cpp
// Method A
return eastl::identical(a.begin(), a.end(), b.begin(), b.end());

// Method B
return (a.size() == b.size()) && eastl::equal(a.begin(), a.end(), b.begin(), b.end());

// Method C 
if (a.size()!= b.size()
    return false;
return eastl::equal(a.begin(), a.end(), b.begin(), b.end());
```

And it seems redundant to separately getting size variables in current implement:

```cpp
const auto sizeA = a.size();
const auto sizeB = b.size();
```

They are just used once.

Reply:
`eastl::identical` is equivalent to `std::equal(first1, last1, first2, last2)`, ie. overloads 5-8 listed at cppref [here](https://en.cppreference.com/w/cpp/algorithm/equal). Method A and B have the same complexity from a Big O notation perspective. Please use method A.

Reply:
@jhopkins-ea Thank you for sharing this information with me.

I agree that method B and C have the same Big O complexity. The difference between them lies solely in the number of times the instructions are executed. and I trust the conditional branch should eventually be optimized away by the compiler. 

Let's talking more about the method A, https://github.com/electronicarts/EASTL/blob/61c6a018e87d2fa1fde43f1ede813f15e8758dfb/include/EASTL/algorithm.h#L1960-L1969

Assume that here are two ring buffer `rb1` (with size 1'000'000) and `rb2` (with size 1'000'010):

```cpp
RBVectorInt rb1 {0, ..., 0} // all elements are 0
RBVectorInt rb2 {0, ..., 0, 1, 0, 0, ..., 0,} // the 1'000'001th element is 1, others are 0
```

In this case, `eastl::identical` need at least 1'000'000 times comparation to get result. But if we can get size information in advance, we can avoid redundant comparation.

Reply:
Let's see what cppref says in [equal](https://en.cppreference.com/w/cpp/algorithm/equal):

> Complexity
> 5,7) At most min(last1 - first1, last2 - first2) applications of the predicate.
However, if InputIt1 and InputIt2 meet the requirements of [LegacyRandomAccessIterator](https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator) and last1 - first1 != last2 - first2 then no applications of the predicate are made (size mismatch is detected without looking at any elements).

See GNU libstdc++ [code](https://github.com/gcc-mirror/gcc/blob/c496d15954cdeab7f9039328f94a6f62cf893d5f/libstdc%2B%2B-v3/include/bits/stl_algobase.h#L1184), they use a `_BoolType` to check whether inputs are RandomAccessIterator or not to optimize specific case.

I guess that's why they don't implement a separate method called `identical`. The interface `identical` currently provided will allow me to form a good habit of comparing the size before calling.

Reply:
> eastl::identical is equivalent to std::equal(first1, last1, first2, last2), ie. overloads 5-8 listed at cppref [here](https://en.cppreference.com/w/cpp/algorithm/equal).

I just realized EASTL doesn't have the `equal` overloads taking 2 iterator pairs. Also, given that `eastl::identical` doesn't do the range size optimization for random access iterators it's technically not the same.

I think I'd prefer us to add those overloads (with the correct optimizations for random access iterators) than to proliferate the usage of `eastl::identical` (which arguably should be deprecated in favor of the new `equal` overloads). But this is out of the scope of this PR.

Given the current state of things, option B is probably the best we can do.

Reply:
> I just realized EASTL doesn't have the `equal` overloads taking 2 iterator pairs. Also, given that `eastl::identical` doesn't do the range size optimization for random access iterators it's technically not the same.

I can provide more information:

When I first find the `eastl::identical` interface, I didn't directly realize that it is an overload of the `equal` function that accepts different parameters. I subconsciously think that this is a "more efficient" `equal`, and the user must ensure that the length is equal before calling. for better performance.

Thanks to @jhopkins-ea for reminding me to establish right perception. XD


> I think I'd prefer us to add those overloads (with the correct optimizations for random access iterators) than to proliferate the usage of `eastl::identical` (which arguably should be deprecated in favor of the new `equal` overloads). But this is out of the scope of this PR.

It would be great if this could be done, I tried to write `eastl::equal(a.begin(), a.end(), b.begin(), b.end())` at first but found that it could not be compiled , that feels strange, because it can be done in other STL implementations, and it will be a better experience to be consistent.

> Given the current state of things, option B is probably the best we can do.

Glad to hear that, are there other suggestions? I think it's time for merging.

Reply:
Good pick up on the complexity differences between `identical` and `equal`. We should in the future deprecate identical in favour of new equal overloads that have the complexity guarantees.

This is a minor issue, but `container.size()`/`std::size()` does not have a guaranteed constant time, as opposed to `ranges::size()` ([see here](https://en.cppreference.com/w/cpp/ranges/size)). If we used a `eastl::list` as the backing container this would be a linear time operation, because `eastl::list` does not store the size (unlike a `std::list`).

Thanks for your contribution.